cmd/cephcsi.go

```diff
@@ -81,6 +81,13 @@ func init() {
 		"",
 		"list of kubernetes node labels, that determines the topology"+
 			" domain the node belongs to, separated by ','")
+	flag.BoolVar(&conf.EnableReadAffinity, "enable-read-affinity", false, "enable read affinity")
+	flag.StringVar(
```

Comment:
> Please add a document that setting this will use the same affinity to all the ceph clusters that cephcsi is connected.

I think that is implied. There's no statement informing users otherwise.

Reply:
Having a note doesn't hurt and it helps users to understand the impact of it.

Reply:
A single Ceph-CSI deployment supports multiple Ceph clusters... What is the reason to have this as a commandline parameter, and not a setting in the StorageClass or the ConfigMap?

Reply:
> A single Ceph-CSI deployment supports multiple Ceph clusters... What is the reason to have this as a commandline parameter, and not a setting in the StorageClass or the ConfigMap?

Its simpler to have it as comandline parameter.
In most of the cases, clusters will share the same crush map characteristics and this design
should be sufficient.

StorageClasses are immutable, therefore existing rbd volumes will not be able to use this feature.

The current implementation covers most of the cases and there should be no need to have this in
a configmap with unique labels per cluster. If such a need arises in the future, it will be easy to
override cmdline labels and prefer labels per cluster in configmap. 

One more reason is of performance, using cmdline args will enable us to restart the container and
fetch the labels only at the time of restart instead of having to fetch it for each map/mount requests.

Reply:
> Its simpler to have it as comandline parameter.

For who? You as a developer, or administrators deploying Ceph-CSI? I'd guess a StorageClass+ConfigMap option would be much easier for administrators.

I have no idea if CRUSH maps are commonly the same on different clusters. My assumption would be that it is highly unlikely, and they would be different per cluster.

IMHO, putting this change in the StorageClass would be suitable, as a SC already refers to a particular cluster.

In case of users that upgrade and want to benefit from this feature, placing it in the ConfigMap should work too?

Fetching labels of nodes needs to be dynamic anyway, I guess? What if a cluster changes, nodes get removed, added, relocated? Autoscaling a Kubernetes cluster is a thing, so fetching the labels only at startup of the provisioner isn't correct.

Reply:
> > Its simpler to have it as comandline parameter.
> 
> For who? You as a developer, or administrators deploying Ceph-CSI? I'd guess a StorageClass+ConfigMap option would be much easier for administrators.

Both, I'd like to think of this feature as set and forget.
Topology of cluster with existing nodes is rarely changed. OSDs are assinged using topology labels on the node.
rbd volumes use the same (with cephcsi's help) to figure out their nearest OSD for read affinity.

> 
> I have no idea if CRUSH maps are commonly the same on different clusters. My assumption would be that it is highly unlikely, and they would be different per cluster.

https://docs.ceph.com/en/quincy/rados/operations/crush-map/#types-and-buckets

We really only care about the CRUSH map type and value, not the exact structure.
 If the cluster is situated in kubernetes, then 
OSD are assigned according to topology labels of the node. This is the same set of topology that will be used by cephcsi to extract the crush location and pass it on to the device while mapping.

CRUSH maps will be not be very different from each other among ceph clusters running on the same kubernetes cluster.

> 
> IMHO, putting this change in the StorageClass would be suitable, as a SC already refers to a particular cluster.

Existing pvc/storageclasses will not be able to take advantage of this feature. 

> 
> In case of users that upgrade and want to benefit from this feature, placing it in the ConfigMap should work too?
> 
> Fetching labels of nodes needs to be dynamic anyway, I guess? What if a cluster changes, nodes get removed, added, relocated? Autoscaling a Kubernetes cluster is a thing, so fetching the labels only at startup of the provisioner isn't correct.

This is happening in the nodeplugin pod and will be used during mapping.
In any of the scenario described above, cephcsi nodeplugin pod will get restarted and we will fetch it again. 
Each nodeplugin pod cares only about its node's labels.

Reply:
I guess we can always move it to the ConfigMap if more flexibility is required.

Right, of course it is only used in the node-plugin, so changes to the number of nodes in the Kubernetes cluster is handled automatically.